### PR TITLE
fix(native-filters): Setting default value

### PR DESCRIPTION
**Pull request by agatapst**
### SUMMARY
The aim of this PR was to improve setting default value for Native Filters. 
The new Default Value Select component was created, which enables to choose default value based on dataset and field.
Now it is possible to set default value for a given filter.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
**Before:**
default_values_before

**After:
for native filter without children (including editing)**
defalult_value_normal_edit

**for native filters child**
default_value_children

### TEST PLAN
Go to `config.py` and set  `"DASHBOARD_NATIVE_FILTERS": True`
Go to dashboards, choose one and create filter or use existing one.
Set default value.

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes one of the issues mentioned in #12148
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc @junlincc @villebro @rusackas 
@adam-stasiak could you test it?

### CODE REVIEW

**Comment by adam-stasiak** [comment]
I have a bug. Scenario:
Go to dashboard and add native filter. Set default value.
Export this dashboard
Import this dashboard
Open imported dashboard
Check default value imported - > should be the same as the original one
Open filter settings and change default filter value
You can see that default filter value will remain the same.

https://user-images.githubusercontent.com/25153919/104360974-23016b80-5512-11eb-9ef2-8f803954163e.mov

**Comment by rusackas** [comment]
> I have a bug. Scenario:
> Go to dashboard and add native filter. Set default value.
> Export this dashboard
> Import this dashboard
> Open imported dashboard
> Check default value imported - > should be the same as the original one
> Open filter settings and change default filter value
> You can see that default filter value will remain the same.
> 
>  import.mov

Whoa... that's some impressive testing!

**Comment by agatapst** [comment]
@adam-stasiak thanks for thorough testing - as always! 🚀 
I think there is a bigger bug with imported dashboards - it is not possible to edit them on master as well. I suggest to separate it - I have created issue for it #12495 that should solve the bug you have described. 

There is a gif presenting that if problem with editing imported dashboard is fixed, we won't have that bug.
imported_dashboard_bug

**Comment by agatapst** [comment]
To avoid issue similar to #12519, I have added resetting default value select for removed dataset.

**Review by suddjian — APPROVED** [review]
awesome work!

**Review by junlincc — COMMENTED** [review]
Great work! 
Next step, let's make it possible to set multiple default values in the modal, reflect the selection in the filter bar dropdown. 

@agatapst 🙏

this one is good to go. ✅

**Comment by adam-stasiak** [comment]
Double checked this - is ok 🟢

**Comment by junlincc** [comment]
@agatapst could we resolve conflicts? we probably should have merge this before the other ones🤦🏾‍♀️
thanks! 🙏

**Comment by agatapst** [comment]
Conflicts solved, but there were some significant changes to adjust this PR to current master, so please wait with merging - I'll check it one more time. 🟡

File: superset-frontend/src/dashboard/components/nativeFilters/DefaultValueSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      {
+        name: ['filters', filterId, 'defaultValue'],
+        touched: false,
+        value: null,
+      },
+    ]);
+  }, [form, filterId]);
+
+  useChangeEffect(datasetId, previous => {
+    if (previous != null) {
```
Comment (simchaNielsen) [inline superset-frontend/src/dashboard/components/nativeFilters/DefaultValueSelect.tsx:60]: may better use strict comparison like: `===` to process correctly `0` or empty string?

File: superset-frontend/src/dashboard/components/nativeFilters/DefaultValueSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      resetFieldValue();
+    }
+  });
+
+  useChangeEffect(column, previous => {
+    if (previous != null) {
```
Comment (simchaNielsen) [inline superset-frontend/src/dashboard/components/nativeFilters/DefaultValueSelect.tsx:66]: and same here

**Comment by simchaNielsen** [comment]
Hi it looks good work 👍 
I have some question about implementation. I see you use `Select` to choose defaultValue but we have different native filters chart types, for example for now we have already implemented `Range` native filter, so I think in this case we can't use `Select` to select appropriate value.

So I think it can be better use SuperChart here to inject appropriate value selection for default value, I had some PR for it but closed it in favour of yours, because you also implemented it in different way... but I still think that it can be good to use SuperChart here to select default value:
https://github.com/apache/superset/pull/12199
cc @villebro thoughts?

**Review by simcha90 — CHANGES_REQUESTED** [review]
@junlincc @suddjian @rusackas just wanted to ask what other devs think about `SuperChart` proposal?
Like here: https://github.com/apache/superset/pull/12199

**Comment by agatapst** [comment]
@simchaNielsen Hi, thanks for your suggestions! Yes, you are right - we use Select component for setting default values for all cases. So for range filters it may not work properly. Tbh creating that PR I didn't know that type should be supported, so thanks for checking that.
Your solution looks nice - I don't mind closing this PR in favor of yours. My PR is completely based on `DefaultValueSelect` component, which in case of using SuperChart, would require to change the whole implementation here.
@villebro @junlincc @rusackas @suddjian I am waiting for your opinion as well.

**Comment by villebro** [comment]
I agree with @simchaNielsen - we should decouple the filters as much as possible from the filter config modal, so it would be preferable to use `SuperChart` there similarly to how it is being used in the filter tab.

**Comment by junlincc** [comment]
@simchaNielsen thanks for the solution, I do agree we should go this route ♥️. I wonder why you closed it earlier though? 

@agatapst thanks for your great work still:) 🙏

**Comment by suddjian** [comment]
Agreed that the other approach makes more sense. Closing this for now.

**Pull request closed without merging**